File: src/NodeParser/SpreadElementNodeParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { ArrayType } from "../Type/ArrayType.js";
import type { InferType } from "../Type/InferType.js";
import type { TupleType } from "../Type/TupleType.js";
import { RestType } from "../Type/RestType.js";

/**
 * Handles `...expr` inside an ArrayLiteralExpression.
 * Turns it into RestType so TupleTypeFormatter can emit correct JSON-Schema.
 */
export class SpreadElementNodeParser implements SubNodeParser {
    constructor(private readonly childNodeParser: NodeParser) {}

    supportsNode(node: ts.SpreadElement): boolean {
        return node.kind === ts.SyntaxKind.SpreadElement;
    }

    createType(node: ts.SpreadElement, context: Context) {
        const inner = this.childNodeParser.createType(node.expression, context) as ArrayType | InferType | TupleType;

        return new RestType(inner);
    }
}

File: internal/parser/spread_element_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// SpreadElementNodeParser handles `...expr` inside an ArrayLiteralExpression,
// turning it into a RestType so the tuple formatter can emit correct JSON
// Schema (src/NodeParser/SpreadElementNodeParser.ts).
type SpreadElementNodeParser struct {
	childNodeParser NodeParser
}

func NewSpreadElementNodeParser(childNodeParser NodeParser) *SpreadElementNodeParser {
	return &SpreadElementNodeParser{childNodeParser: childNodeParser}
}

func (p *SpreadElementNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindSpreadElement
}

func (p *SpreadElementNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	return &types.RestType{Type: p.childNodeParser.CreateType(node.AsSpreadElement().Expression, ctx, nil)}
}
